libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/mpp/syncKotlinAndAndroidSourceSets.kt

```diff
@@ -0,0 +1,145 @@
+        -- could be claimed by Android androidTest (instrumented test)
+
+    The Kotlin source set would win in this scenario.
+     */
+    if (disambiguationClassifier == null || !androidSourceSet.name.startsWith(disambiguationClassifier)) {
+        kotlinSourceSet.kotlin.srcDir("src/${androidSourceSet.name}/kotlin")
+    }
+
+    kotlinSourceSet.kotlin.srcDirs(*androidSourceSet.java.srcDirs.toTypedArray())
+    androidSourceSet.java.srcDirs(*kotlinSourceSet.kotlin.srcDirs.toTypedArray())
+
+    /*
+    Make sure to include user configuration as well.
+    Unfortunately, there does not exist any "ad-hoc" API like `all`.
+    Therefore we sync the directories once again after evaluation
+     */
+    target.project.whenEvaluated {
+        kotlinSourceSet.kotlin.srcDirs(*androidSourceSet.java.srcDirs.toTypedArray())
+        androidSourceSet.java.srcDirs(*kotlinSourceSet.kotlin.srcDirs.toTypedArray())
+    }
+}
+
+private fun syncKotlinAndAndroidResources(
+    target: KotlinAndroidTarget,
+    kotlinSourceSet: KotlinSourceSet,
+    androidSourceSet: AndroidSourceSet
+) {
+    /*
+    Non-MPP projects will not have a different name for kotlin source sets vs Android source sets
+    Registering resource folders is unnecessary therefore.
+     */
+    if (kotlinSourceSet.name == androidSourceSet.name) {
+        return
+    }
+
+    val project = target.project
+
+    androidSourceSet.resources.srcDirs(*kotlinSourceSet.resources.toList().toTypedArray())
+    if (androidSourceSet.resources.srcDirs.isNotEmpty()) {
+        androidSourceSet.resources.srcDir(kotlinSourceSet.sourceFolderFor(project, "resources"))
+        kotlinSourceSet.resources.srcDirs(androidSourceSet.resources.srcDirs)
+    }
+
+    if (androidSourceSet.assets.srcDirs.isNotEmpty()) {
+        androidSourceSet.assets.srcDir(kotlinSourceSet.sourceFolderFor(project, "assets"))
+    }
+
+    if (androidSourceSet.res.srcDirs.isNotEmpty()) {
+        androidSourceSet.res.srcDir(kotlinSourceSet.sourceFolderFor(project, "res"))
+    }
+
+    if (androidSourceSet.aidl.srcDirs.isNotEmpty()) {
+        androidSourceSet.aidl.srcDir(kotlinSourceSet.sourceFolderFor(project, "aidl"))
+    }
+
+    if (androidSourceSet.renderscript.srcDirs.isNotEmpty()) {
+        androidSourceSet.renderscript.srcDir(kotlinSourceSet.sourceFolderFor(project, "rs"))
+    }
+
+    if (androidSourceSet.jni.srcDirs.isNotEmpty()) {
+        androidSourceSet.jni.srcDir(kotlinSourceSet.sourceFolderFor(project, "jni"))
+    }
+
+    if (androidSourceSet.jniLibs.srcDirs.isNotEmpty()) {
+        androidSourceSet.jniLibs.srcDir(kotlinSourceSet.sourceFolderFor(project, "jniLibs"))
+    }
+
+    if (androidSourceSet.shaders.srcDirs.isNotEmpty()) {
+        androidSourceSet.shaders.srcDir(kotlinSourceSet.sourceFolderFor(project, "shaders"))
+    }
+}
+
+private fun KotlinSourceSet.sourceFolderFor(project: Project, type: String): File {
+    return project.file("src/${this.name}/$type")
```

Comment:
Maybe avoid duplication with that function? https://github.com/JetBrains/kotlin/blob/bd66e87440b9d2463437a4682128d4ce24de91ca/libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/sources/KotlinSourceSetFactory.kt#L33. This is a very simple piece of logic, though, and finding the duplicate will be easy if we need to change it.

Reply:
Good point! See here https://github.com/JetBrains/kotlin/pull/3918/commits/b580114dd70a28b8d36297b42d86789ce579149a
✅